### PR TITLE
Field Definition mit optionalem `Location` Argument

### PR DESCRIPTION
**Pull request by marcometz**
graphql Calls umgebaut:

```
query { categories{
  name
  genericItemsCount(location: "Hamburg")
  genericItems(location: "Berlin"){title}
  pointsOfInterestCount
  pointsOfInterest{name}
  eventRecordsCount
  toursCount
  tours{name}
  upcomingEventRecordsCount
  upcomingEventRecords{title}
  }
}
```

Bildschirmfoto 2022-04-27 um 15 05 11

SVA-475

### CODE REVIEW

**Comment by donni106** [comment]
ℹ️ follow up for #301

**Comment by donni106** [comment]
Thanks, I tested with the following query:

```
query Categories($location: String) {
  categories {
    id
    name
    genericItemsCount(location: $location)
    genericItems(location: $location) {
      title
    }
    eventRecordsCount(location: $location)
    upcomingEventRecordsCount(location: $location)
    pointsOfInterestCount(location: $location)
    toursCount(location: $location)
    pointsOfInterest(location: $location) {
      location {
        name
        regionName
        state
      }
    }
  }
}
```

 compared to the query based on the base branch:

```query Categories($location: String) {
  categories(location: $location) {
    id
    name
    genericItemsCount
    genericItems {
      title
    }
    eventRecordsCount
    upcomingEventRecordsCount
    pointsOfInterestCount
    toursCount
    pointsOfInterest {
      location {
        name
        regionName
        state
      }
    }
  }
}
```

the results are the same except that most of the `genericItems` are `null` now instead of `[]`. 

Bildschirmfoto 2022-04-28 um 11 45 53

this should always be an empty array to avoid type differences across all results.

File: app/models/static_content.rb [hunk 1]
```diff
@@ -34,7 +34,7 @@ def self.sorted_and_filtered_for_params(params)
   private
 
     def enforce_nil_for_empty_version
-      self.version = nil if version.empty?
+      self.version = nil if version.try(:empty?)
```
Comment (donni106) [inline app/models/static_content.rb:37]: why is this addressed in this pull request?

File: app/models/data_resources/resource_modules/category.rb [hunk 2]
```diff
@@ -15,45 +15,27 @@ class Category < ApplicationRecord
...
     event_records.upcoming(nil)
   end
 
-  def upcoming_event_records_count
-    return 0 if event_records.blank?
-    return 0 if upcoming_event_records.blank?
+  # Defines 2 methods for a list of record type:
+  # - Returns the number of items the given location and record type
+  # - Returns the items of the given location and record type
+  [:event_records, :upcoming_event_records, :generic_items, :points_of_interest, :tours].each do |item_source|
```
Comment (donni106) [inline app/models/data_resources/resource_modules/category.rb:25]: `upcoming_event_records` was forgotten to be removed, right? and on the other hand, the `upcoming` scope is missing in the generated methods i guess?

File: app/models/data_resources/resource_modules/category.rb [hunk 3]
```diff
@@ -15,45 +15,27 @@ class Category < ApplicationRecord
...
-    return 0 if tours.blank?
-
-    tours.visible.count
-  end
-
-  def news_items_count
-    return 0 if news_items.blank?
-
-    news_items.visible.count
-  end
```
Comment (donni106) [inline app/models/data_resources/resource_modules/category.rb:40]: `news_items_count` is entirely removed but not listed in the symbols for generating methods.

File: app/models/data_resources/resource_modules/category.rb [hunk 4]
```diff
@@ -15,45 +15,27 @@ class Category < ApplicationRecord
...
   end
 
-  def upcoming_event_records_count
-    return 0 if event_records.blank?
-    return 0 if upcoming_event_records.blank?
+  # Defines 2 methods for a list of record type:
+  # - Returns the number of items the given location and record type
+  # - Returns the items of the given location and record type
+  [:event_records, :upcoming_event_records, :generic_items, :points_of_interest, :tours].each do |item_source|
+    define_method "#{item_source}_by_location" do |args = nil|
```
Comment (donni106) [inline app/models/data_resources/resource_modules/category.rb:26]: why are the methods now `_by_location`? i would remove that like before. the methods can receive an optional param `location` which then does a filtering, but with it, the results are not "by location".

File: app/models/data_resources/resource_modules/category.rb (hunk 2 above)
Comment (donni106) [reply app/models/data_resources/resource_modules/category.rb:25]: ah ok the method `upcoming_event_records` is still needed to be used by the generated method for upcoming records, as this is not a default model.

File: app/models/data_resources/resource_modules/category.rb [hunk 5]
```diff
@@ -15,45 +15,27 @@ class Category < ApplicationRecord
...
+      return nil if send(item_source).blank?
```
Comment (donni106) [inline app/models/data_resources/resource_modules/category.rb:27]: the default return should be `[]` right? see my example posted in the conversation before.

File: app/models/static_content.rb (hunk 1 above)
Comment (marcometz) [reply app/models/static_content.rb:37]: i stumbled across this when trying to test this feature with an empty db

File: app/models/data_resources/resource_modules/category.rb (hunk 2 above)
Comment (marcometz) [reply app/models/data_resources/resource_modules/category.rb:25]: exactly

File: app/models/data_resources/resource_modules/category.rb (hunk 3 above)
Comment (marcometz) [reply app/models/data_resources/resource_modules/category.rb:40]: i even had not news_items in my graphql test.
so i will readd the method as before!

File: app/models/data_resources/resource_modules/category.rb (hunk 4 above)
Comment (marcometz) [reply app/models/data_resources/resource_modules/category.rb:26]: i need another method name to deviate from the has_many relation.
by the has_many relation (: points_of_interest, ...), there are already getter-methods defined which i did not want to overwrite in general/all cases.

File: app/models/data_resources/resource_modules/category.rb (hunk 3 above)
Comment (marcometz) [reply app/models/data_resources/resource_modules/category.rb]: fixed by readding lost method

**Review by donni106 — APPROVED** [review]

**Pull request merged**